### PR TITLE
Create Workato Connect adapter

### PR DESCRIPTION
**Pull request by ivan-pencic-devtech**
Implement triggers:
	Poll fulfillment requests
Implement actions:
	Purchase, Change quantity and Cancel request

### CODE REVIEW

File: cloudblue_connect.custom_adapter.rb [hunk 1]
```diff
@@ -0,0 +1,694 @@
+#
+# This file is a part of the Ingram Micro’s CloudBlue Connect Workato Extension.
+
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete this line. It must be just five lines of copyright
\#
\# This file is a part of the Ingram Micro’s CloudBlue Connect Workato Extension.
\#
\# @copyright (c) 2021. Ingram Micro. All Rights Reserved.
\#

File: cloudblue_connect.custom_adapter.rb [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+#
+
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete this line

File: cloudblue_connect.custom_adapter.rb [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+# @copyright (c) 2021. Ingram Micro. All Rights Reserved.
+
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete this line

File: cloudblue_connect.custom_adapter.rb [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+#
+
+
+{
+  title: "CloudBlue Connect",
+
+  # API key authentication example. See more examples at https://docs.workato.com/developing-connectors/sdk/guides/authentication.html
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete this comment, it's from Workato sample

File: cloudblue_connect.custom_adapter.rb [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+        label: "API Token",
+        hint: "You can find your API Token in <a href='https://portal.connect.cloudblue.com/integrations/tokens/' target='_blank'>Connect</a>",
+        optional: false
+      },
+    ],
+
+    authorization: {
+      type: "custom_auth",
+
+      # Calendly uses header auth (Authorization: <api key>)
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete this comment, it has no relation to the connector, it's from Workato sample

File: cloudblue_connect.custom_adapter.rb [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+            { name: "external_id"},
+            { name: "external_uid"}
+            ]
+          },
+          { name: "errors", type: "array"},
+          
+          
+         ]
+      end
+    },
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: these structures look very similar. suggestion is to replace them with a single structure 'created_request'

File: cloudblue_connect.custom_adapter.rb [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+  methodes: {
+  },
+  actions: {
+
+
+    asset_cancel_request: {
+      title: "Create Asset Cancel Request",
+      subtitle: "",
+      description: "",
+      help: "<a href='https://connect.cloudblue.com/community/api/openapi/#operations-Assets-request_list_listRequests' target='_blank'>Official documentation</a>",
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: this looks like incorrect URL. Correct URL is https://connect.cloudblue.com/community/api/openapi/#operations-Assets-request_list_createRequest

File: cloudblue_connect.custom_adapter.rb [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+            error("#{message}: #{body}")
+          end
+      end,
+      
+      output_fields: lambda do |object_definitions|
+        object_definitions["cancel_request"]
+      end
+    },
+    
+    asset_purchase_request: {
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb:195]: this looks like incorrect URL. Correct URL is https://connect.cloudblue.com/community/api/openapi/#operations-Assets-request_list_createRequest

File: cloudblue_connect.custom_adapter.rb [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+      title: "Create Asset Purchase Request",
+      subtitle: "",
+      description: "",
+      help: "<a href='https://connect.cloudblue.com/community/api/openapi/#operations-Assets-request_list_listRequests' target='_blank'>Official documentation</a>",
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: this looks like incorrect URL. Correct URL is https://connect.cloudblue.com/community/api/openapi/#operations-Assets-request_list_createRequest

File: cloudblue_connect.custom_adapter.rb [hunk 10]
```diff
@@ -0,0 +1,694 @@
...
+        [
+          {
+            name: "external_id",
+            label: "Asset External ID",
+            hint: "Provide the ID of the asset from the external system, for example pointing to a contract identifier or unique one that will map this asset with.",
+            optional: true
+          },
+          {
+            name: "external_uid",
+            label: "Asset External UID1",
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: why is it UID1? why '1' in the end?

File: cloudblue_connect.custom_adapter.rb [hunk 11]
```diff
@@ -0,0 +1,694 @@
...
+                    {
+            name: "tier1_city",
+            label: "Tier 1 City",
+            hint: "Please specify the City name of the related account, please note that are not validated by CloudBlue Connect but vendor may do it.
+
+
+
+",
+            optional: false
+          },
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: formatting is broken, please fix

File: cloudblue_connect.custom_adapter.rb [hunk 12]
```diff
@@ -0,0 +1,694 @@
...
+          
+          {
+            name: "tier1_country",
+            label: "Tier 1 Country",
+            hint: "Provide the country of the related account, please use Alpha-2 code.",
+            optional: false
+          },
+         {
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: indent

File: cloudblue_connect.custom_adapter.rb [hunk 13]
```diff
@@ -0,0 +1,694 @@
...
+            optional: false
+          },
+          {
+            name: "tier1_email",
+            label: "Tier 1 Email",
+            hint: "Provide the email address of the technical contact related with the account, this one typically is a requirement by vendors to process the request, additionally please note that this email will be used to send inquire requests related with the fulfillment of the request.
+",
+            optional: false
+          },
+         {
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: indent

File: cloudblue_connect.custom_adapter.rb [hunk 14]
```diff
@@ -0,0 +1,694 @@
...
+            optional: false
+          },
+          
+              {
+            name: "customer_contact_first_name",
+            label: "Customer Contact First Name",
+            hint: "Provide the contact name of the technical contact related with the account.",
+            optional: false
+          },
+
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: formatting, indents

File: cloudblue_connect.custom_adapter.rb [hunk 15]
```diff
@@ -0,0 +1,694 @@
...
+            label: "Product Items",
+            hint: "",
+            optional: false,
+            type: "array", of: "object", properties: [
+            { optional: false, name: "id", label: "Item ID" },
+            { optional: false, name: "quantity", label: "Quantity" }
+            ]
+          },
+          
+
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: formatting indents

File: cloudblue_connect.custom_adapter.rb [hunk 16]
```diff
@@ -0,0 +1,694 @@
...
+      end,
+      
+      execute: lambda do |connection, input|
+        
+        x = 4
+        rand_uid_num = (rand(10 ** x) + 10**x).to_s[0..x-1]
+        rand_uid_num_two = (rand(10 ** x) + 10**x).to_s[0..x-1]
+        rand_id_num = (rand(10 ** x) + 10**x).to_s[0..x-1]
+        first_item = input["product_items"][0]
+        #"id"=>"#{first_item["id"].slice(0, first_item["id"].rindex("-"))}"
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: why do we need it? there is "workato.uuid" function to generate UID

File: cloudblue_connect.custom_adapter.rb [hunk 17]
```diff
@@ -0,0 +1,694 @@
...
+           "type"=>"purchase",
+           "asset"=>{
+           "product"=>{
+           "id"=>"#{input["product_id"]}"
+             
+      },
+      "connection"=>{
+         "id"=>"CT-0000-0000-0000",
+         "type"=>"production"
+      },
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: I think we should omit this. Connect will derive connection from product and hub/marketplace.

File: cloudblue_connect.custom_adapter.rb [hunk 18]
```diff
@@ -0,0 +1,694 @@
...
+                  }
+               }
+            }
+           
+         }
+      }
+   },
+   "marketplace"=>{
+      "id"=>"MP-90040"
+   }
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: Marketplace ID should be optional input field, not a hardcoded field.
When empty, Connect will derive it from hub ID and tier1 / tier2 contact - there is a configuration in Connect for this.

File: cloudblue_connect.custom_adapter.rb [hunk 19]
```diff
@@ -0,0 +1,694 @@
...
+        
+        params = {
+          "type": "cancel",
+          "asset": {
+            "external_id": "#{input["external_id"]}",
+            "external_uid": "#{input["external_uid"]}"
+          }
+        }
+   
+        post("https://api.connect.cloudblue.com/public/v1/requests", params).
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: it should be 
```suggestion
        post("/public/v1/requests", params).
```
workato to construct the initial part of URL using base-uri method

File: cloudblue_connect.custom_adapter.rb [hunk 20]
```diff
@@ -0,0 +1,694 @@
...
+            }
+           
+         }
+      }
+   },
+   "marketplace"=>{
+      "id"=>"MP-90040"
+   }
+  }  
+        post("https://api.connect.cloudblue.com/public/v1/requests", hardcoded_params).
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: it should be 
```suggestion
        post("/public/v1/requests", hardcoded_params).
```
workato wil construct the rest using base_uri method

File: cloudblue_connect.custom_adapter.rb [hunk 21]
```diff
@@ -0,0 +1,694 @@
...
+      end
+    },
+        
+    
+    
+    change_items_quantity: {
+      title: "Change items quantity",
+      subtitle: "",
+      description: "",
+      help: "",
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: For sake of consistency, this action must be "Asset change request"
Also need to add help link

File: cloudblue_connect.custom_adapter.rb [hunk 22]
```diff
@@ -0,0 +1,694 @@
...
+      
+              #"global_id": input["items"],
+              #"quantity": input["quantity"],
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete

File: cloudblue_connect.custom_adapter.rb [hunk 23]
```diff
@@ -0,0 +1,694 @@
...
+        response = post("/public/v1/requests", payload).
+        after_error_response(/.*/) do |_, body, _, message|
+          parse_json(body)
+        end
+        #post("/public/v1/requests", payload).
+        #after_error_response(/.*/) do |_, body, _, message|
+        #  {
+        #    data: body
+        #  }
+        #end
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete

File: cloudblue_connect.custom_adapter.rb [hunk 24]
```diff
@@ -0,0 +1,694 @@
...
+    # Picklists can be referenced by inputs fields or object_definitions
+    # possible arguements - connection
+    # see more at https://docs.workato.com/developing-connectors/sdk/sdk-reference/picklists.html
+
+  },
+
+  # Reusable methods can be called from object_definitions, picklists or actions
+  # See more at https://docs.workato.com/developing-connectors/sdk/sdk-reference/methods.html
+  methods: {
+  }
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: delete.
there already was definition of 'methods' above

**Review by elikhtanskiy — CHANGES_REQUESTED** [review]
comments inline

File: cloudblue_connect.custom_adapter.rb [hunk 25]
```diff
@@ -0,0 +1,659 @@
...
+
+                      ]
+                    end,
+
+      execute: lambda do |connection, input|
+        customer_external_uid = if input['customer_external_uid'].blank?
+                                  workato.uuid
+                                else
+                                  call(:get_tier_external_uid, connection, input['customer_external_id'])
+                                end
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: Why is it this algorithm?

Correct algorithm seems to be
```suggestion
        customer_external_uid = if input['customer_external_uid'].blank?
                                  #if no UID in input, try to look for it by external ID (or autogenerate if not found)
                                  call(:get_tier_external_uid, connection, input['customer_external_id']) 
                                else
                                  #if UID is passed in the input explicitly, let's use it
                                  input['customer_external_uid']
                                end
```

File: cloudblue_connect.custom_adapter.rb [hunk 26]
```diff
@@ -0,0 +1,659 @@
...
+                               workato.uuid
+                             else
+                               input['asset_external_uid']
+                             end
+
+        tier1_external_uid = if input['tier1_external_uid'].blank?
+                               workato.uuid
+                             else
+                               call(:get_tier_external_uid, connection, input['tier1_external_uid'])
+                             end
```
Comment (elikhtanskiy) [inline cloudblue_connect.custom_adapter.rb]: Same here

```suggestion
        tier1_external_uid = if input['tier1_external_uid'].blank?
                               call(:get_tier_external_uid, connection, input['tier1_external_uid'])
                             else
                               input['tier1_external_uid']
                             end
```

**Review by elikhtanskiy — CHANGES_REQUESTED** [review]

File: cloudblue_connect.custom_adapter.rb (hunk 25 above)
Comment (ivan-pencic-devtech) [reply cloudblue_connect.custom_adapter.rb]: Yes that was a mistake, I made another commit that will fix this

File: cloudblue_connect.custom_adapter.rb (hunk 26 above)
Comment (ivan-pencic-devtech) [reply cloudblue_connect.custom_adapter.rb]: Same mistake as above, fix is pushed

**Pull request merged**